### DISCUSSION TITLE
Custom error message for email:dns validation rule

### DISCUSSION THREAD

**Discussion by SuperDJ** [laravel/framework#42038] [Q&A]
I use the below validation rule for validating emails. The default error message of "The :attribute must be a valid email address." is fine in most cases but I think it sometimes lacks information. For example for the end user it would be clearer to know that the domain is invalid if the dns rule fails.
```php
'email:rfc,strict,dns,spoof'
```

Is it possible to use a custom error message like below?
```php
'email.email:dns' => ':attribute domain name invalid'
```

**Comment by Kamandlou** [comment]
Yes it is.

If you use $this->validate() simplest one, then you should write code something like this:

```
$rules = [
        'name' => 'required',
        'email' => 'required|email',
        'message' => 'required|max:250',
    ];

    $customMessages = [
        'required' => 'The :attribute field is required.',
        'email' => 'Custom validation message'
    ];

    $this->validate($request, $rules, $customMessages);
```

**Reply by SuperDJ** [reply]
Yes, I know about custom error messages. But the question is about `email:dns` not just `email` validation

**Reply by rvanbaalen** [reply]
Have you found the solution yet? I'm looking for the same thing.

**Reply by yannisc** [reply]
same question here...

**Comment by hussein4alaa** [comment]
public function rules()
    {
        return [
            'email' => 'email:rfc,dns'
        ];
    }


    public function messages()
    {
        return [
            'email.email' => 'custom message',
        ];
    }

**Comment by meredevelopment** [comment]
Not an answer, but something I tried that didn't cause any errors, but didn't actually work as hoped: 
```
$messages = [
    'email.email' => [
        'dns' => 'Bad :attribute DNS',
        'spoof' => ':attribute might be spoofed,
        'filter' => ':attribute fails ye olde FilterVar check'
    ],
]

$attributes = [
    'email' => 'E-Mail Address',
];

$validator = Validator::make($data, [
    'email' => ['email:filter,spoof,dns']
], $messages, $attributes);
```

Using the above, if any of the validations fail, the error contains _all_ the error messages. So for example testing against an email address using a domain that I know isn't registered, and logging the result with:

```
$errors = $validator->errors();
Log::error($errors);
```

shows this in the log: 
`[2023-05-30 11:21:41] prod.ERROR: {"email":[{"dns":"Bad E-Mail Address DNS","spoof":"E-Mail Address might be spoofed","filter":"E-Mail Address fails ye olde FilterVar check"}]}`

Anyone got a workaround for this?

**Reply by tbalint1995** [reply]
I have reviewed several websites and even spent a lot of time searching, but currently, I couldn't find any information on how to list additional rules from an email in Laravel. At least, I couldn't find any relevant resources on the internet. I have spent two days on this so far.

**Comment by jellebreuer** [comment]
I did this with a workaround, i just created a custom rule that does the validations 1 by 1

```php
<?php

namespace App\Rules;

use Closure;
use Illuminate\Contracts\Validation\ValidationRule;
use Validator;

class EmailRule implements ValidationRule
{
    /**
     * Run the validation rule.
     *
     * @param  \Closure(string): \Illuminate\Translation\PotentiallyTranslatedString  $fail
     */
    public function validate(string $attribute, mixed $value, Closure $fail): void
    {
        $validator = Validator::make([$attribute => $value], [
            $attribute => ['email:rfc'],
        ]);
        if ($validator->fails()) {
            $fail($validator->errors()->first());
        }

        $validator = Validator::make([$attribute => $value], [
            $attribute => ['email:dns'],
        ]);
        if ($validator->fails()) {
            $fail(__('custom message'));
        }
    }
}
```
